src/CSV.jl

```diff
@@ -1,139 +1,266 @@
+    if r.result isa Missing
+        return missing
+    else
+        @inbounds pool = f.pools[col]
+        return get🐱(pool, r.result::Tuple{Ptr{UInt8}, Int})
+    end
+end
+
+@inline function parsefield(f, T, row, col, strict)
+    r = Parsers.parse(f.parsinglayers, f.io, T; f.kwargs...)
+    if !Parsers.ok(r.code)
+        strict ? throw(Error(Parsers.Error(f.io, r), row, col)) :
+            println("warning: failed parsing $T on row=$row, col=$col, error=$(Parsers.codes(r.code))")
+    end
+    return r.result
 end
 
-function Base.show(io::IO, f::Source)
-    println(io, "CSV.Source: ", f.fullpath)
-    println(io, f.io)
-    show(io, f.schema)
+@noinline skipcells(f, n) = foreach(x->Parsers.parse(f.parsinglayers, f.io, Tuple{Ptr{UInt8}, Int}), 1:n)
+
+Base.getproperty(csvrow::Row{F}, name::Symbol) where {F <: File{NT}} where {NT} =
+    getproperty(csvrow, Tables.columntype(NT, name), Tables.columnindex(NT, name), name)
+
+function Base.getproperty(csvrow::Row{F}, ::Type{T}, col::Int, name::Symbol) where {T, F <: File{NT, transpose}} where {NT, transpose}
+    f = getfield(csvrow, 1)
+    row = getfield(csvrow, 2)
+    if transpose
+        @inbounds Parsers.fastseek!(f.io, f.positions[col])
+    else
+        lastparsed = f.lastparsedcol[]
+        if col === lastparsed + 1
+        elseif col === 1
+            @inbounds Parsers.fastseek!(f.io, f.positions[row])
+        elseif col > lastparsed + 1
+            # skipping cells
+            skipcells(f, col - lastparsed+1)
+        elseif col !== lastparsed + 1
+            # randomly seeking within row
+            @inbounds Parsers.fastseek!(f.io, f.positions[row])
+            skipcells(f, col - 1)
+        end
+    end
+    r = parsefield(f, parsingtype(T), row, col, f.strict)
+    if transpose
+        @inbounds f.positions[col] = position(f.io)
+    else
+        f.lastparsedcol[] = col
+    end
+    return r
+end
+
+File(source::Union{String, IO};
+    # file options
+    use_mmap::Bool=true,
+    # header can be a row number, range of rows, or actual string vector
+    header::Union{Integer, UnitRange{Int}, Vector}=1,
+    # by default, data starts immediately after header or start of file
+    datarow::Int=-1,
+    footerskip::Int=0,
+    transpose::Bool=false,
+    # parsing options
+    missingstrings=String[],
+    missingstring="",
+    delim::Union{Char, String}=",",
+    quotechar::Union{UInt8, Char}='"',
+    openquotechar::Union{UInt8, Char, Nothing}=nothing,
+    closequotechar::Union{UInt8, Char, Nothing}=nothing,
+    escapechar::Union{UInt8, Char}='\\',
+    dateformat::Union{String, Dates.DateFormat, Nothing}=nothing,
+    decimal::Union{UInt8, Char, Nothing}=nothing,
+    truestrings::Union{Vector{String}, Nothing}=nothing,
+    falsestrings::Union{Vector{String}, Nothing}=nothing,
+    # type options
+    types=nothing,
+    typemap::Dict=Dict{Type, Type}(),
+    allowmissing::Symbol=:all,
+    categorical::Bool=false,
```

Comment:
If you break the API, it would be a good time to replace `categorical` with `strings=:categorical`, as these two arguments are not really independent. But I see the `strings` argument isn't there anymore. Does that mean one cannot avoid interning strings?

BTW, should add one indentation space to get correct alignment.

Reply:
Yes, strings are interned by default in the Parsers.jl package. I don't see any reason not to always intern strings? And for WeakRefStrings, users can state them specifically in `types` or there's a new `typemap` keyword argument to swap a type for type for the whole file.

Reply:
It's maybe not essential, but it happens that a CSV file contains strings which never appear twice, e.g. names, IDs or free-form answers to a questionnaire. But I guess this would need to be decided on a per-column basis, so maybe we don't need the `strings` argument after all. Another use case would be to choose a custom string, for example to use a special encoding, but that can be added later.

At any rate I'd remove the `length(levels[col]) / sum(values(levels[col])) < .67` hack: better always return categorical or non-categorical when the user passes either `true` or `false`, and only use the trick when the user passes a real (or even remove it).